### PR TITLE
Hyperspy build

### PR DESCRIPTION
**Pull request by to266**
Initial hyperspy build

### CODE REVIEW

**Comment by to266** [comment]
Eh... how do I force it to only build py3?

**Comment by msarahan** [comment]
I don't know why setup.py is not being found by the CI systems.  It does appear to be in the tarball that you specify.  Does this recipe work locally?

**Comment by to266** [comment]
How should I test it..? Sorry, I'm more or less just poking in the dark with this recipe.

Strangely enough [appveyor here](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.2797) finds the setup for some builds, and doesn't for others...

**Comment by msarahan** [comment]
Use conda-build directly on your system:

```
conda install conda-build
conda build <path to your recipe> [--python=X.Y]
```

where you can optionally specify python version, if it differs from your root anaconda/miniconda installation.

File: recipes/hyperspy/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    - hyperspy
+  commands:
+    - python hyperspy/continuous_integration/nosetest.py
+  requires:
+    - mock
+    - nose
+
+about:
+  home: https://hyperspy.org
+  license: GPL
```
Comment (msarahan) [inline recipes/hyperspy/meta.yaml]: What version of the GPL?

**Comment by to266** [comment]
Thanks for the help, @msarahan. I can build it locally, and it generally only seems to fail here when `setuptools` are updated (which in turn downgrades `conda-build` and `conda-build-all`), which seems to trigger the failure.

Any further guidance appreciated :)

**Comment by msarahan** [comment]
You should not need conda-build or conda-build-all as build dependencies.  Only what you need to build your package - those are tools one level up.

The setuptools update coupled with conda-build and conda-build-all sounds like an infrastructure bug on our side.  @jakirkham @ocefpaf have either of you seen anything like this?

**Comment by to266** [comment]
@msarahan any progress with this? I'm willing to help with figuring this out

**Comment by msarahan** [comment]
It looks to me like AppVeyor is only failing on Python 3.4 x64.  The Mac build on Travis starts with Python 3.4, and also fails; not continuing to Python 3.5.

The root install appears to be Python 3.5.  It may be relevant to test Python 3.4 builds with a Python 3.5 root locally.  I'll try that.

**Comment by to266** [comment]
FYI it works for me (both 3.4 and 3.5 builds), however I'm running linux (x64, root 3.5).

As an aside, should I temporarily attempt to skip 3.4 builds?

**Comment by msarahan** [comment]
IMHO, fine to skip 3.4 builds.  I have not been able to duplicate the failures locally with either Win x64 py34 or OS X py34.  I did see one test failure:

```
======================================================================
FAIL: hyperspy.tests.signal.test_eds_sem.Test_metadata.test_take_off_angle
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/msarahan/miniconda3/envs/_test/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/msarahan/miniconda3/envs/_test/lib/python3.4/site-packages/hyperspy/tests/signal/test_eds_sem.py", line 163, in test_take_off_angle
    nose.tools.assert_equal(s.get_take_off_angle(), 12.886929785732487)
AssertionError: 12.886929785732484 != 12.886929785732487
```

You're seeing limits of floating point precision.  You should probably use assert_close or something like that, not assert_equal.

**Comment by to266** [comment]
Yeah, that's fixed in the next version (should be released today or tomorrow or at least this week). I'll give it a go

**Comment by to266** [comment]
Ok, so only travis fails now, and it seems to be the only one with root Python 3.4

(appveyor is 2.7, and circleci 3.5)

**Comment by msarahan** [comment]
I think this might be related to pinning conda build to an old version. @jakirkham what's the latest on that?

**Comment by to266** [comment]
Anything new? Is there a way to specify the `conda-build` version in the recipe (I can see that the requirements did not work)?

**Comment by msarahan** [comment]
There's an example PR where I unpinned it in the CI scripts at https://github.com/conda-forge/staged-recipes/pull/693 - ideally the solution is that we unpin conda-build in the master version of those CI scripts (in staged-recipes HEAD), and you just pull those changes in.  However, that needs discussion from @jakirkham @ocefpaf @pelson

**Comment by to266** [comment]
Can you roughly estimate the timescale for the unpinning? I'm mostly wondering if that will be done before this Tuesday or not.

**Comment by jakirkham** [comment]
So, what is the pinning problem here? AppVeyor appears to be passing. Is it Travis CI that is the issue? Have we tested this?

**Comment by msarahan** [comment]
I tested unpinning in https://github.com/conda-forge/staged-recipes/pull/693

Windows is passing here because @to266 removed the Python 3.4 build completely.

**Comment by jakirkham** [comment]
Ok, go ahead and remove the skip as this has been tested. It will fail, but we will ignore this. It can be patched in the CI script in the feedstock.

Now what's up with the test failure on Travis? Is this just a case of floating point rounding error and a bit too precise test? Could we patch the test suite to skip this brittle test until it can be fixed? Alternatively, we can patch with a fix and submit the fix upstream. Either should be incredibly easy to do. If not, please ask for help.

**Comment by to266** [comment]
The fragile test is (should be) fixed on our (hyperspy) side, so only the conda-build issue on travis is left.

I've re-enabled the 3.4 build.

**Comment by to266** [comment]
Huh, probably should not have merged the upstream. Sorry if there's confusion about that

**Comment by jakirkham** [comment]
So, I noticed that you have `setuptools` as a [setup requirement]( https://github.com/hyperspy/hyperspy/blob/76a8cfb233cccb11116adc26622bfa75cf8e32ca/setup.py#L152 ) in `hyperspy`. I don't really follow how this works, but could this be causing some problems also?

File: recipes/hyperspy/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    - natsort
+    - lxml
+    - ipyparallel
+    - dask
+    - traits >=4.5
+    - traitsui >=5.0
+    - requests
+    - ipywidgets >=5.0
+    - tqdm
+    - pip:
```
Comment (msarahan) [inline recipes/hyperspy/meta.yaml]: I think this is causing the error on the CI's.  You can't use pip here.  If dateutil is not available as a conda package, please create a recipe and add it with your PR here.

Comment (msarahan) [inline recipes/hyperspy/meta.yaml]: Also, I apologize for any confusion.  The yaml here is unfortunately different from the yaml used by conda-env, which does allow pip in syntax like this.

Comment (to266) [inline recipes/hyperspy/meta.yaml]: Could not figure that out, thanks. 

Did I understand correctly, I should add the dateutil recipe to this same PR?

I guessed I'd have to create a new PR..

Comment (msarahan) [inline recipes/hyperspy/meta.yaml]: With only one, and that one being a pure-python lib, it's OK here.  Generally discouraged because it makes debugging take longer.  Just a judgement call.  Could do as another PR.

**Comment by msarahan** [comment]
All green. Merging.

**Comment by jakirkham** [comment]
So, we are kind of backlogged at the moment due to some issues with Travis CI's API ( https://github.com/conda-forge/staged-recipes/issues/923 ). Once they clear this should convert cleanly to a feedstock. We appreciate your patience as we sort this out.
